test/common/upstream/load_balancer_fuzz_base.cc

```diff
@@ -0,0 +1,241 @@
+    priority_of_host_set++;
+  }
+  num_priority_levels_ = priority_of_host_set;
+}
+
+// Updating host sets is shared amongst all the load balancer tests. Since logically, we're just
+// setting the mock priority set to have certain values, and all load balancers interface with host
+// sets and their health statuses, this action maps to all load balancers.
+void LoadBalancerFuzzBase::updateHealthFlagsForAHostSet(const uint64_t host_priority,
+                                                        const uint32_t num_healthy_hosts,
+                                                        const uint32_t num_degraded_hosts,
+                                                        const uint32_t num_excluded_hosts,
+                                                        const std::string random_bytestring) {
+  const uint8_t priority_of_host_set = host_priority % num_priority_levels_;
+  ENVOY_LOG_MISC(trace, "Updating health flags for host set at priority: {}", priority_of_host_set);
+  MockHostSet& host_set = *priority_set_.getMockHostSet(priority_of_host_set);
+  // This downcast will not overflow because size is capped by port numbers
+  uint32_t host_set_size = host_set.hosts_.size();
+  host_set.healthy_hosts_.clear();
+  host_set.degraded_hosts_.clear();
+  host_set.excluded_hosts_.clear();
+
+  Fuzz::ProperSubsetSelector subset_selector(random_bytestring);
+
+  std::vector<std::vector<uint8_t>> subsets = subset_selector.constructSubsets(
+      {num_healthy_hosts, num_degraded_hosts, num_excluded_hosts}, host_set_size);
+
+  // Healthy hosts are first subset
+  for (uint8_t index : subsets.at(0)) {
+    host_set.healthy_hosts_.push_back(host_set.hosts_[index]);
+    ENVOY_LOG_MISC(trace, "Index of host made healthy at priority level {}: {}",
+                   priority_of_host_set, index);
+  }
+
+  // Degraded hosts are second subset
+  for (uint8_t index : subsets.at(1)) {
+    host_set.degraded_hosts_.push_back(host_set.hosts_[index]);
+    ENVOY_LOG_MISC(trace, "Index of host made degraded at priority level {}: {}",
+                   priority_of_host_set, index);
+  }
+
+  // Excluded hosts are third subset
+  for (uint8_t index : subsets.at(2)) {
+    host_set.excluded_hosts_.push_back(host_set.hosts_[index]);
+    ENVOY_LOG_MISC(trace, "Index of host made excluded at priority level {}: {}",
+                   priority_of_host_set, index);
+  }
+
+  // Handle updating health flags for hosts_per_locality_
+  HostVector healthy_hosts_locality_one = {};
+  HostVector degraded_hosts_locality_one = {};
+  HostVector excluded_hosts_locality_one = {};
+
+  HostVector healthy_hosts_locality_two = {};
+  HostVector degraded_hosts_locality_two = {};
+  HostVector excluded_hosts_locality_two = {};
+
+  HostVector healthy_hosts_locality_three = {};
+  HostVector degraded_hosts_locality_three = {};
+  HostVector excluded_hosts_locality_three = {};
+
+  for (uint8_t index : subsets.at(0)) {
```

Comment:
This loop and the two that follow, look pretty similar.
Do you think it would be possible to generalize this to a single nested loop (iterating over the subsets indices)?

Reply:
The major problem is the three HostVectors acted upon in each loop are different. Do you see any other way of modularizing this? (perhaps a function?) I agree, it would be nice to generalize this.

Reply:
There are 2 things that I think can be done:
1. you may be able to define `std::array<HostVector, 3> per_locality_healthy_hosts`, etc., and use the `locality_indexes_[index]` value to decide which HostVector the host should be pushed into.
2. The above arrays can be wrapped in another array, and the index for that array should be matched with the index of `subsets`.

Reply:
Awesome, thanks for the idea! Drawing this out now.

Reply:
Finished refactor. Made code a lot more concise. Thanks so much!